### PR TITLE
feat(drizzle): Migrate to Drizzle ORM

### PR DESCRIPTION
**Pull request by realmikesolo**
This PR was made to migrate from [Prisma](https://www.prisma.io/) to [Drizzle ORM](https://orm.drizzle.team/docs/overview). It involves rewriting the database interactions using `Drizzle ORM`.

# Instructions

## If you directly start the project with Drizzle ORM

1. **Remove unnecessary schema file**: Delete `lib/legacy-schema.ts` file as it is only used when migrating from Prisma.
2. **Initialize Schema**: The Drizzle schema located in `lib/schema.ts` will be used for database queries.
2. **Apply changes to the database**: Run the `drizzle-kit push` command to apply your changes to the database. Learn more about the push command [here](https://orm.drizzle.team/kit-docs/overview#prototyping-with-db-push).
3. **Begin using the template**: You can now start using this template with Drizzle ORM.

## If you migrating from Prisma

1. **Replace the schema file**: Remove the existing `lib/schema.ts` file and rename `lib/legacy-schema.ts` to `lib/schema.ts`.
2. **Update database schema**: `email` column in `users` table is set to `not null` to ensure compatibility with drizzle next-auth adapter. Apply this change by running the `drizzle-kit push` command. Learn more about `push` command [here](https://orm.drizzle.team/kit-docs/overview#prototyping-with-db-push).
3. **Complete migration**: You are now ready to use this template with Drizzle ORM.

### CODE REVIEW

File: app/[domain]/[slug]/page.tsx [hunk 1]
```diff
@@ -113,7 +109,7 @@ export default async function SitePostPage({
           href={
             data.site?.user?.username
               ? `https://twitter.com/${data.site.user.username}`
-              : `https://github.com/${data.site?.user?.gh_username}`
+              : `https://github.com/${data.site?.user?.ghUsername}`
```
Comment (leerob) [inline app/[domain]/[slug]/page.tsx:112]: Would this require a data migration?

File: lib/actions.ts [hunk 2]
```diff
@@ -33,18 +32,17 @@ export const createSite = async (formData: FormData) => {
...
-            id: session.user.id,
-          },
-        },
-      },
-    });
+        userId: session.user.id,
+      })
+      .returning();
+
+    // unnecessary await
```
Comment (leerob) [inline lib/actions.ts:45]: Remove the await?

File: lib/actions.ts [hunk 3]
```diff
@@ -202,92 +190,97 @@ export const updateSite = withSiteAuth(
...
-});
+export const deleteSite = withSiteAuth(
+  async (_: FormData, site: SelectSite) => {
+    try {
+      const [response] = await db
+        .delete(sites)
+        .where(eq(sites.id, site.id))
+        .returning();
+
+      await revalidateTag(
```
Comment (leerob) [inline lib/actions.ts:201]:
```suggestion
      revalidateTag(
```

File: lib/actions.ts [hunk 4]
```diff
@@ -202,92 +190,97 @@ export const updateSite = withSiteAuth(
...
+        `${site.subdomain}.${process.env.NEXT_PUBLIC_ROOT_DOMAIN}-metadata`,
+      );
+      response.customDomain &&
+        (await revalidateTag(`${site.customDomain}-metadata`));
```
Comment (leerob) [inline lib/actions.ts:205]:
```suggestion
        revalidateTag(`${site.customDomain}-metadata`);
```

File: lib/actions.ts [hunk 5]
```diff
@@ -202,92 +190,97 @@ export const updateSite = withSiteAuth(
...
-  site.customDomain && (await revalidateTag(`${site.customDomain}-posts`));
+    const [response] = await db
+      .insert(posts)
+      .values({
+        siteId: site.id,
+        userId: session.user.id,
+      })
+      .returning();
+
+    await revalidateTag(
```
Comment (leerob) [inline lib/actions.ts:243]:
```suggestion
    revalidateTag(
```

**Review by leerob — COMMENTED** [review]
This is looking great!

**Comment by leerob** [comment]
@m-torin if you would like to use Prisma, that's great. Please do! I don't have anything against Prisma. In fact, we just worked with them on this https://db-latency.vercel.app/.

**Comment by realmikesolo** [comment]
I made some edits based on the review, updated the `Drizzle` schema, upgraded `drizzle-kit` to version `0.21.0`, and created a new `Prisma` compatible schema with an initial migration. These can be used if switching from `Prisma` to `Drizzle ORM`. The migration was created after pulling the DDL from an existing database, which was initially used with `Prisma`. The only thing that needs to be done is to set the `email` field in the `User` table to 'not null' for type compatibility with the `Drizzle` adapter for `next-auth`.

Currently, there are two bugs related to `drizzle-kit` removing foreign key constraints and recreating them, and `drizzle-kit` detecting changes in `timestamp` precision even when there are none. These should be fixed soon. Once they are resolved, the `Prisma` compatible schema can be used. If it's a new project, it's recommended to use the standard `Drizzle` schema.

Please let me know if anything needs updating.

File: app/[domain]/[slug]/page.tsx [hunk 6]
```diff
@@ -1,10 +1,12 @@
 import { notFound } from "next/navigation";
-import prisma from "@/lib/prisma";
 import { getPostData, getSiteData } from "@/lib/fetchers";
 import BlogCard from "@/components/blog-card";
 import BlurImage from "@/components/blur-image";
 import MDX from "@/components/mdx";
 import { placeholderBlurhash, toDateString } from "@/lib/utils";
+import db from "@/lib/db/db";
```
Comment (AlexBlokh) [inline app/[domain]/[slug]/page.tsx:7]: inline

File: components/posts.tsx [hunk 7]
```diff
@@ -15,23 +18,26 @@ export default async function Posts({
...
-      updatedAt: "desc",
-    },
-    include: {
-      site: true,
-    },
-    ...(limit ? { take: limit } : {}),
-  });
 
-  return posts.length > 0 ? (
+  const query = db
```
Comment (AlexBlokh) [inline components/posts.tsx:22]: make in query syntax

File: components/sites.tsx [hunk 8]
```diff
@@ -1,29 +1,33 @@
...
-      },
-    },
-    orderBy: {
-      createdAt: "asc",
-    },
-    ...(limit ? { take: limit } : {}),
-  });
+  const query = db
+    .select({
+      ...getTableColumns(sites),
```
Comment (AlexBlokh) [inline components/sites.tsx:17]: redundant

**Comment by realmikesolo** [comment]
Bugs with Drizzle-kit were fixed

**Review by leerob — APPROVED** [review]
Thank you!

**Comment by mdichtler** [comment]
@leerob  @realmikesolo - this seems to break Github auth as it might happen that Github login that's in the example returns null for the email, due to the schema email field being set to notNull in users table. This happens when the person trying to login in Github email preferences has Keep my email addresses private checked.

**Comment by mdichtler** [comment]
Submitted issue for tracking: https://github.com/vercel/platforms/issues/412

**Comment by mdichtler** [comment]
@AlexBlokh - I'm free for the next hour if you want to troubleshoot this a bit - add my discord: @martindichtler also attaching screenshot of my github profile where email is set to null. This is logged from auth.ts line 14 (profile callback)

image

**Comment by AlexBlokh** [comment]
@mdichtler here's a [conversation](https://discord.com/channels/1200116961590399008/1231534836859146241/1252233393224290314) with the NextAuth team in Discord 

They advise us to do a migration of `id -> email` where email is null before declaring a schema and seems like they consider this a bug, but we tend to just make this field `nullable` since it doesn't make sense to have it `not null`

**Pull request merged**